notes/shelley/Interface.md

```diff
@@ -0,0 +1,461 @@
+logic parts.
+
+There will be a single thread in the application layer which clears the
+input queue from the diffusion layer and forks a thread with a handler for
+each.
+
+```Haskell
+data InputFromDiffusionLayer =
+    BlockHeader HeaderHash
+  | Transaction Tx
+  | MPC RelevantMPCData
+  | Update RelevantUpdateData
+  | ...
+
+type Application m = InputFromDiffusionLayer -> m ()
+
+type InputQueue = TBQueue InputFromDiffusionLayer
+
+withDiffusionLayerInput :: Application m -> InputQueue -> m x
+withDiffusionLayerInput go q = do
+  inp <- readTBQueue q
+  _ <- async (go inp)
+  withDiffusionLayerInput go q
+```
+
+### Inv/Req/Data
+
+Much of the network interaction is handled by the Inv/Req/Data relay system.
+Pulling this down into the diffusion layer should be straightforward. It will
+use synchronous `diffusion => application` API functions to get the data in
+response to requests, and the asynchronous API in response to data from peers
+(the data will be passed on through some queue).
+
+### Block, block headers, and recovery mode
+
+Block and block header communication and relaying does not use the Inv/Req/Data
+system.
+
+#### Block header announcement and block serving
+
+To announce a block header, simply pass it to the diffusion layer. The
+implementation of serving blocks and block headers will be moved to the
+diffusion layer, and supported by the synchronous API to get blocks and block
+headers (from the database, which is owned by the application layer). In
+future we'll put in a diffusion layer cache, but that's not part of this
+initial work.
+
+#### Block retrieval
+
+Block retrieval has more moving parts than header announcement. There's a
+retrieval queue which came to be because block headers are sent without
+their bodies, and the recipient is not offered a chance to ask for the body
+within the same conversation. So the headers are thrown into a queue and some
+thread will eventually request the body (unless it considered useless by the
+time it comes out of the queue).
+
+We can leave this in place for the initial work, and discuss a better solution
+later. The application layer handler for block headers will remain largely the
+same, except that it will take input from the diffusion layer and drop it into
+the application layer queue as it does now. The block retrieval worker will
+pick it up and do the relevant logic, unless...
+
+#### Recovery mode
+
+Sometimes the block retrieval worker will trigger recovery mode, in case it
+finds a header which doesn't immediately continue the tip. This recovery
+process uses time-warp conversations for control flow: it will get the headers
+and blocks needed to go from current tip to new tip, and then return to the
+block retrieval queue clearing loop, possibly ending recovery mode.
+
+We can keep this as is (equally bad, not worse) by using the synchronous
+`getBlocks` function and letting the diffusion layer take care of finding them.
+
+## Concrete details of implementation
```

Comment:
From here down is some actual Haskell text of how I think we should form this. There's also `Interface.hs` which shows a working example without any actual functionality. I'm ready to run with this but obviously I'd like some feedback first.

@dcoutts @georgeee